autoload/vista/parser/lsp.vim

```diff
@@ -41,80 +43,103 @@ function! s:IsFileUri(uri) abort
+    \ }
+endfunction
+
+function! s:LocalToRawSymbol(sym)
+  return {
+    \ 'line': a:sym.lnum,
+    \ 'kind': a:sym.kind,
+    \ 'name': a:sym.text,
+    \ }
+endfunction
+
+function! s:IsDocumentSymbol(sym)
+  return has_key(a:sym, 'selectionRange')
+endfunction
+
+function! s:ParseSymbolInfoList(outlist, symbols) abort
+  for lspsym in a:symbols
+    let l:loc = lspsym.location
+    if s:IsFileUri(l:loc.uri)
+      call add(a:outlist, s:LspToLocalSymbol(lspsym, l:loc.range))
     endif
-  endif
+  endfor
 endfunction
 
-function! vista#parser#lsp#CocSymbols(symbol, container) abort
-  if vista#ShouldIgnore(a:symbol.kind)
-    return
-  endif
+function! s:ParseDocumentSymbolsRec(outlist, symbols, level) abort
+  for lspsym in a:symbols
+    let l:sym = s:LspToLocalSymbol(lspsym, lspsym.selectionRange)
+    let l:sym.level = a:level
+    call add(a:outlist, l:sym)
+    if has_key(lspsym, 'children')
+      call s:ParseDocumentSymbolsRec(a:outlist, lspsym.children, a:level + 1)
+    endif
+  endfor
+endfunction
 
-  let raw = { 'line': a:symbol.lnum, 'kind': a:symbol.kind, 'name': a:symbol.text }
-  call add(g:vista.raw, raw)
+function! s:GroupSymbolsByKind(symbols) abort
+  let l:groups = {}
+  for l:sym in a:symbols
+    let l:list = get(l:groups, l:sym.kind)
+    if !has_key(l:groups, l:sym.kind)
+      let l:list = []
+      let l:groups[l:sym.kind] = l:list
+    endif
+    call add(l:list, l:sym)
+  endfor
+  return l:groups
+endfunction
 
-  if a:symbol.kind ==? 'Method' || a:symbol.kind ==? 'Function'
-    call add(g:vista.functions, a:symbol)
+function! vista#parser#lsp#ParseDocumentSymbolPayload(resp) abort
+  let l:symbols = []
+  if s:IsDocumentSymbol(a:resp[0])
+    call s:ParseDocumentSymbolsRec(l:symbols, a:resp, 0)
+    call vista#parser#lsp#FilterDocumentSymbols(l:symbols)
+    call vista#parser#lsp#DispatchDocumentSymbols(l:symbols)
+    return l:symbols
+  else
+    call s:ParseSymbolInfoList(l:symbols, a:resp)
+    call vista#parser#lsp#FilterDocumentSymbols(l:symbols)
+    return s:GroupSymbolsByKind(l:symbols)
   endif
+endfunction
 
-  call add(a:container, {
-        \ 'lnum': a:symbol.lnum,
-        \ 'col': a:symbol.col,
-        \ 'text': a:symbol.text,
-        \ 'kind': a:symbol.kind,
-        \ 'level': a:symbol.level
-        \ })
+function! vista#parser#lsp#FilterDocumentSymbols(symbols) abort
+  let l:symlist = a:symbols
+  if exists('g:vista_ignore_kinds')
+    call filter(l:symlist, 'index(g:vista_ignore_kinds, v:val) < 0')
+  endif
+  let g:vista.functions = []
+  for l:sym in l:symlist
+    if l:sym.kind ==? 'Method' || l:sym.kind ==? 'Function'
+      call add(g:vista.functions, l:sym)
+    endif
+  endfor
+  return l:symlist
 endfunction
 
-" https://microsoft.github.io/language-server-protocol/specification#textDocument_documentSymbol
-function! vista#parser#lsp#ExtractSymbol(symbol, container) abort
-  let symbol = a:symbol
+function! vista#parser#lsp#DispatchDocumentSymbols(symbols)
```

Comment:
This seems to be used only once, therefore it's not worthy to define another function.